### PR TITLE
rgw: Make rgw admin ops api get user info consistent with the command line

### PR DESCRIPTION
**Pull request by letterwuyu**
GET /{admin}/user?format=json HTTP/1.1
Host: {fqdn}
This api gets the information is incomplete relative to radosgw-admin user info --uid xxxx

This modification will change the information returned by the three calls :
RGWUserAdminOp_User::info
RGWUserAdminOp_User::create
RGWUserAdminOp_User::modify

Signed-off-by: Li Shuhao <lishuhao@unitedstack.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by votdev** [comment]
Are these changes fully tested via existing QA unit tests? Does it break the Dashboard RGW tests?

**Comment by cbodley** [comment]
jenkins test make check

**Comment by votdev** [comment]
Please enhance the RGW user API tests in https://github.com/ceph/ceph/blob/master/qa/tasks/radosgw_admin_rest.py, too. Some tests are missing, e.g. checking for relevant fields, like tenant, in the returned data.

**Comment by cbodley** [comment]
thanks for the review @votdev. are those api tests only run by the mgr suite? it may be useful to add some subset of those to the rgw suite (which is all we usually run) so we don't break things

i'm not sure how useful the tenant field is in the reply, since you have to include the tenant name as part of the request itself. but it'd be fine to add a field for it in RGWUserInfo::dump()

**Comment by votdev** [comment]
> thanks for the review @votdev. are those api tests only run by the mgr suite? it may be useful to add some subset of those to the rgw suite (which is all we usually run) so we don't break things

Yes, the Dashboard related RGW API tests are only executed by the mgr suite.

> i'm not sure how useful the tenant field is in the reply, since you have to include the tenant name as part of the request itself. but it'd be fine to add a field for it in RGWUserInfo::dump()

We need the ``tenant`` field in the output because of https://tracker.ceph.com/issues/36480 and https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/controllers/rgw.py#L156. Some of the RGW Admin OPS endpoints require a uid like ``[<tenant>$]<user>``.

**Comment by votdev** [comment]
@cbodley See also https://github.com/ceph/ceph/pull/24757#pullrequestreview-168482764 :-)

**Comment by letterwuyu** [comment]
Hi，Sorry for replying so late, I have added the test, please help me run teuthology.

**Comment by ivancich** [comment]
Please address review requests. And let's not add needs-qa until the reviews are clean.

**Comment by letterwuyu** [comment]
> The `tenant` field is missing in the output. This breaks the [Dashboard RGW API QA tests](https://github.com/ceph/ceph/blob/master/qa/tasks/mgr/dashboard/test_rgw.py#L327).
hi @votdev  I have change code

File: src/rgw/rgw_user.cc [hunk 1]
```diff
@@ -726,25 +726,53 @@ static void dump_user_info(Formatter *f, RGWUserInfo &info,
...
   dump_access_keys_info(f, info);
   dump_swift_keys_info(f, info);
-  info.caps.dump(f);
+
+  encode_json("caps", info.caps, f);
+
+  char buf[256];
+  op_type_to_str(info.op_mask, buf, sizeof(buf));
+  encode_json("op_mask", (const char *)buf, f);
+  encode_json("system", (bool)system, f);
```
Comment (votdev) [inline src/rgw/rgw_user.cc]: Where does this variable coming from? I assume this should be
```suggestion
  encode_json("system", (bool)info.system, f);
```

**Comment by votdev** [comment]
Could you please rebase your PR, otherwise it is necessary to manually fix multiple merge conflicts.

File: src/rgw/rgw_user.cc (hunk 1 above)
Comment (letterwuyu) [reply src/rgw/rgw_user.cc]: Oh, thx, this is my fault.

**Comment by letterwuyu** [comment]
@votdev  thx，I have changed code and rebase master

**Comment by votdev** [comment]
Please review this PR @theanalyst, @cbodley, @tchaikov, @mattbenjamin.

**Comment by cbodley** [comment]
```
2019-03-26T21:37:41.161 INFO:teuthology.run_tasks:Running task radosgw-admin-rest...
2019-03-26T21:37:41.176 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 86, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 64, in run_one_task
    task = get_task(taskname)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 28, in get_task
    module = _import('tasks', module_name, task_name, fail_on_import_error=True)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 52, in _import
    0,
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/builtins.py", line 93, in __import__
    result = _import(*args, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-cbodley-testing/qa/tasks/radosgw_admin_rest.py", line 241
    assert out['caps'] = []
                       ^
SyntaxError: invalid syntax
```
from http://qa-proxy.ceph.com/teuthology/cbodley-2019-03-26_20:23:04-rgw-wip-cbodley-testing-distro-basic-smithi/3775745/teuthology.log

File: qa/tasks/radosgw_admin_rest.py [hunk 2]
```diff
@@ -247,6 +266,26 @@ def task(ctx, config):
     assert out['keys'][0]['access_key'] == access_key
     assert out['keys'][0]['secret_key'] == secret_key
     assert not out['suspended']
+    assert out['tenant'] == ''
+    assert out['max_buckets'] == 4
+    assert out['caps'] = []
+    assert out['op_mask'] = "read, write, delete"
```
Comment (votdev) [inline qa/tasks/radosgw_admin_rest.py]:
```suggestion
    assert out['op_mask'] == "read, write, delete"
```

File: qa/tasks/radosgw_admin_rest.py [hunk 3]
```diff
@@ -247,6 +266,26 @@ def task(ctx, config):
     assert out['keys'][0]['access_key'] == access_key
     assert out['keys'][0]['secret_key'] == secret_key
     assert not out['suspended']
+    assert out['tenant'] == ''
+    assert out['max_buckets'] == 4
+    assert out['caps'] = []
```
Comment (votdev) [inline qa/tasks/radosgw_admin_rest.py]:
```suggestion
    assert out['caps'] == []
```

File: qa/tasks/radosgw_admin_rest.py [hunk 4]
```diff
@@ -247,6 +266,26 @@ def task(ctx, config):
...
+    assert out['op_mask'] = "read, write, delete"
+    assert out['default_placement'] == ''
+    assert out['default_storage_class'] == ''
+    assert out['placement_tags'] = []
```
Comment (votdev) [inline qa/tasks/radosgw_admin_rest.py]:
```suggestion
    assert out['placement_tags'] == []
```

**Comment by letterwuyu** [comment]
@cbodley  @votdev  Hi, thx, I have fix this code. please review and rerun teuthology

**Comment by ivancich** [comment]
@letterwuyu There's no linked tracker. Does this need backports? If so, please add a tracker and link things up.

**Comment by letterwuyu** [comment]
@ivancich Hi
http://tracker.ceph.com/issues/39134
http://tracker.ceph.com/issues/39135

**Comment by smithfarm** [comment]
@letterwuyu The main issue should have been opened in the "rgw" project, not "Stable Releases". Fixing.

**Pull request merged**